File: src/main/java/jp/aoichaan0513/A_TosoGame_Live/Listeners/onInventory.java
package jp.aoichaan0513.A_TosoGame_Live.Listeners;

import jp.aoichaan0513.A_TosoGame_Live.API.MainAPI;
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.Inventory.Left.PlayerListInventory;
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.Inventory.MainInventory;
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.Inventory.Right.MissionInventory;
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.MissionManager;
import jp.aoichaan0513.A_TosoGame_Live.API.Scoreboard.Teams;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.event.Event;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.event.inventory.InventoryOpenEvent;
import org.bukkit.event.inventory.InventoryType;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.InventoryView;
import org.bukkit.inventory.ItemStack;

public class onInventory implements Listener {

    public static boolean isAllowOpen = false;

    @EventHandler
    public void onInventoryClick(InventoryClickEvent e) {
        Player p = (Player) e.getWhoClicked();
        Inventory inventory = e.getInventory();
        ItemStack itemStack = e.getCurrentItem();

        if (!Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_ADMIN, p)) {
            if (Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_PLAYER, p) || Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_SUCCESS, p) || Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_JAIL, p)) {
                if (inventory.getType() == InventoryType.PLAYER) {
                    if (itemStack == null || itemStack.getType() == Material.AIR) return;
                    if (itemStack.getType() == Material.MAP || itemStack.getType() == Material.FILLED_MAP) {
                        e.setResult(Event.Result.DENY);
                        e.setCancelled(true);
                    }
                } else if (inventory.getType() == InventoryType.CHEST || inventory.getType() == InventoryType.ENDER_CHEST
                        || inventory.getType() == InventoryType.DISPENSER || inventory.getType() == InventoryType.DROPPER
                        || inventory.getType() == InventoryType.HOPPER || inventory.getType() == InventoryType.ANVIL
                        || inventory.getType() == InventoryType.SHULKER_BOX || inventory.getType() == InventoryType.BARREL) {
                    if (!MissionManager.isMission()) {
                        e.setResult(Event.Result.DENY);
                        e.setCancelled(true);
                    }
                }
            } else if (Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_HUNTER, p) || Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_TUHO, p)) {
                e.setResult(Event.Result.DENY);
                e.setCancelled(true);
            }
        }
    }

    @EventHandler
    public void onInventoryOpen(InventoryOpenEvent e) {
        Player p = (Player) e.getPlayer();
        Inventory inventory = e.getInventory();
        InventoryView inventoryView = e.getView();

        if (!Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_ADMIN, p)) {
            if (Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_PLAYER, p) || Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_SUCCESS, p) || Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_JAIL, p)) {
                if (inventory.getType() == InventoryType.CHEST || inventory.getType() == InventoryType.ENDER_CHEST
                        || inventory.getType() == InventoryType.DISPENSER || inventory.getType() == InventoryType.DROPPER
                        || inventory.getType() == InventoryType.HOPPER || inventory.getType() == InventoryType.ANVIL
                        || inventory.getType() == InventoryType.SHULKER_BOX || inventory.getType() == InventoryType.BARREL
                        || inventory.getType() == InventoryType.BREWING || inventory.getType() == InventoryType.FURNACE) {
                    if (inventoryView.getTitle().equals(MainInventory.title) || inventoryView.getTitle().equals(PlayerListInventory.title)
                            || inventoryView.getTitle().equals(MissionInventory.missionTitle) || inventoryView.getTitle().equals(MissionInventory.tutatuHintTitle)
                            || inventoryView.getTitle().equals(MissionInventory.endTitle))
                        return;

                    if (MissionManager.isMission() || isAllowOpen) return;
                    e.setCancelled(true);
                    p.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "現在開くことが出来ません。");
                }
            } else if (Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_HUNTER, p) || Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_TUHO, p)) {
                if (inventory.getType() == InventoryType.CHEST || inventory.getType() == InventoryType.ENDER_CHEST
                        || inventory.getType() == InventoryType.DISPENSER || inventory.getType() == InventoryType.DROPPER
                        || inventory.getType() == InventoryType.HOPPER || inventory.getType() == InventoryType.ANVIL
                        || inventory.getType() == InventoryType.SHULKER_BOX) {
                    if (inventoryView.getTitle().equals(MainInventory.title) || inventoryView.getTitle().equals(PlayerListInventory.title)
                            || inventoryView.getTitle().equals(MissionInventory.missionTitle) || inventoryView.getTitle().equals(MissionInventory.tutatuHintTitle)
                            || inventoryView.getTitle().equals(MissionInventory.endTitle))
                        return;

                    e.setCancelled(true);
                    p.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "現在開くことが出来ません。");
                }
            }
        }
    }
}

File: src/main/kotlin/jp/aoichaan0513/A_TosoGame_Live/Listeners/Minecraft/onInventory.kt
package jp.aoichaan0513.A_TosoGame_Live.Listeners.Minecraft

import jp.aoichaan0513.A_TosoGame_Live.API.MainAPI
import jp.aoichaan0513.A_TosoGame_Live.API.MainAPI.PrefixType
import jp.aoichaan0513.A_TosoGame_Live.Inventory.*
import jp.aoichaan0513.A_TosoGame_Live.Mission.MissionManager
import jp.aoichaan0513.A_TosoGame_Live.Utils.isAdminTeam
import jp.aoichaan0513.A_TosoGame_Live.Utils.isHunterGroup
import jp.aoichaan0513.A_TosoGame_Live.Utils.isJailTeam
import jp.aoichaan0513.A_TosoGame_Live.Utils.isPlayerGroup
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.event.Event
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.inventory.*
import org.bukkit.inventory.InventoryView

class onInventory : Listener {

    companion object {
        var isAllowOpen = false
    }

    private val ITEM_BLACKS = setOf(
            Material.FILLED_MAP, Material.BOOK,
            Material.BONE, Material.FEATHER,
            Material.EGG, Material.SNOWBALL,
            Material.RABBIT_FOOT, Material.SLIME_BALL,
            Material.ENDER_PEARL, Material.ENDER_EYE
    )

    @EventHandler
    fun onInventoryClick(e: InventoryClickEvent) {
        val p = e.whoClicked as Player
        val clickedInventory = e.clickedInventory
        // val inventory = e.inventory
        val inventoryView = e.view
        val itemStack = e.currentItem
        val slotType = e.slotType

        if (p.isAdminTeam) return

        if (p.isPlayerGroup || p.isJailTeam) {
            if (slotType == InventoryType.SlotType.CRAFTING) {
                e.result = Event.Result.DENY
                e.isCancelled = true
            } else {
                if (clickedInventory?.type != InventoryType.PLAYER) return

                if (itemStack == null || itemStack.type == Material.AIR || isAllowedInventory(inventoryView)) return

                if (ITEM_BLACKS.contains(itemStack.type)) {
                    e.result = Event.Result.DENY
                    e.isCancelled = true
                }
            }
        } else if (p.isHunterGroup) {
            e.result = Event.Result.DENY
            e.isCancelled = true
        }
    }

    @EventHandler
    fun onInventoryOpen(e: InventoryOpenEvent) {
        val p = e.player as Player
        // val inventory = e.inventory
        val inventoryView = e.view

        if (p.isAdminTeam) return

        if (isAllowedInventory(inventoryView)) return
        if (p.isPlayerGroup || p.isJailTeam) {
            if (MissionManager.isInventoryAllowOpenMission || isAllowOpen) return

            e.isCancelled = true
            p.sendMessage("${MainAPI.getPrefix(PrefixType.ERROR)}現在開くことができません。")
        } else if (p.isHunterGroup) {
            e.isCancelled = true
            p.sendMessage("${MainAPI.getPrefix(PrefixType.ERROR)}現在開くことができません。")
        }
    }

    @EventHandler
    fun onInventoryDrag(e: InventoryDragEvent) {
        val slots = e.rawSlots

        if (slots.any { it in 1..4 }) {
            e.result = Event.Result.DENY
            e.isCancelled = true
        }
    }

    @EventHandler
    fun onCraftItem(e: CraftItemEvent) {
        val p = e.whoClicked as Player

        if (p.isAdminTeam) return

        e.isCancelled = true
    }

    private fun isAllowedInventory(inventoryView: InventoryView): Boolean {
        val title = inventoryView.title
        return title == MainInventory.title || title == MissionInventory.missionTitle
                || title == MissionInventory.tutatuHintTitle || title == MissionInventory.endTitle
                || title == PlayerSettingsInventory.title || title == PlayerSettingsInventory.inventoryTitle
                || title == PlayerSettingsInventory.itemSelectTitle || title.startsWith(CallInventory.title)
                || title.startsWith(ResultInventory.title)
    }
}